packages/query-core/src/queryClient.ts

```diff
@@ -202,41 +141,28 @@ export class QueryClient {
     updater: Updater<TQueryFnData | undefined, TQueryFnData | undefined>,
     options?: SetDataOptions,
   ): TQueryFnData | undefined {
-    const query = this.queryCache.find<TQueryFnData>(queryKey)
+    const query = this.queryCache.find<TQueryFnData>({ queryKey })
     const prevData = query?.state.data
     const data = functionalUpdate(updater, prevData)
 
     if (typeof data === 'undefined') {
       return undefined
     }
 
-    const parsedOptions = parseQueryArgs(queryKey)
-    const defaultedOptions = this.defaultQueryOptions(parsedOptions)
+    const defaultedOptions = this.defaultQueryOptions({ queryKey })
     return this.queryCache
       .build(this, defaultedOptions)
-      .setData(data, { ...options, manual: true })
+      .setData(data, { ...options, manual: true }) as TQueryFnData
```

Comment:
why is this change needed?

Reply:
for a reason, without the `as TQueryFnData`, typescript reads that `setData` is returning `unknown`

![without-as-TQueryFnData](https://user-images.githubusercontent.com/36894846/209797757-91e0eff9-4370-4239-a04a-887dc658e0ef.png)

While inside `setData` it infers the first argument type and returns it so it should return the `TQueryFnData` 

![setData](https://user-images.githubusercontent.com/36894846/209797859-9527e9f1-ef65-4845-9889-de928de6a861.png)

but for a reason that is not happening when we use setData inside setQueryData, although the type of `data` passed to the `setData` is `TQueryFnData` not `unknown`

![typof data passed to setData](https://user-images.githubusercontent.com/36894846/209798770-ad1ed681-ff3f-4ac8-bd31-76a8c4834c2b.png)

Reply:
ah yes, so would:

```
.setData<TQueryFnData>(data, { ...options, manual: true })
```

work as well?

Reply:
Fixed ✅ 

I traced it and tried to find the difference between my branch and the main branch 

In the main branch:
- queryKey was passed to parseQueryArgs which returns a type of `QueryOptions<any, any, any, QueryKey>`
- when the `parsedOptions` is passed to `defaultQueryOptions` it returns ` DefaultedQueryObserverOptions<any, any, unknown, any, QueryKey>`

![main-branch](https://user-images.githubusercontent.com/36894846/209826911-e25c2d6b-237d-4293-a1e8-310b6e144eeb.png)

but in my branch, because I was passing the `queryKey` only to `defaultQueryOptions` it inferred the type to be `DefaultedQueryObserverOptions<unknown, unknown, unknown, unknown, QueryKey>`

![my-branch-before](https://user-images.githubusercontent.com/36894846/209827270-dc5897b6-4b0a-4f2b-9635-6eac31e6112d.png)

so to fix it instead of `as` alias, I passed the generics to the `defaultQueryOptions` by myself

![my-branch after](https://user-images.githubusercontent.com/36894846/209827571-846cb8a1-c82a-48a1-9989-91c1efb5dd6a.png)